src/App.tsx

```diff
@@ -1,24 +1,198 @@
+  const [inputValue, setInputValue] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [deletingId, setDeletingId] = useState<number[]>([]);
+
+  const visibeTodos = filteredTodos(todos, filter);
+  const activeTodosQuantity = activeTodosCount(todos).length;
+  const completedTodos = visibeTodos.length - activeTodosQuantity;
+  const completedTodoList = filteredTodos(todos, Filter.Completed);
+
+  useEffect(() => {
+    const loadTodos = async () => {
+      try {
+        const loadedTodos = await getTodos();
+
+        setTodos(loadedTodos);
+      } catch (err) {
+        throw new Error();
+      } finally {
+        setIsSelected(false);
+      }
+    };
+
+    loadTodos();
+  }, []);
+
+  const setNotification = (message: Errors) => {
+    setErrorNotification(message);
+    setError(true);
+  };
+
+  const clearNotification = useCallback(() => {
+    setError(false);
+  }, []);
+
+  const clearInput = () => {
+    setInputValue('');
+  };
+
+  const handleCreateTodo = useCallback((title: string) => {
+    clearNotification();
+
+    if (!title) {
+      setNotification(Errors.Empty);
+    }
+
+    if (title.trim()) {
+      const newTodo = {
+        id: 0,
+        title,
+        userId: USER_ID,
+        completed: false,
+      };
+
+      setTempTodo(newTodo);
+      addTodo(newTodo)
+        .then((todo) => {
+          setTodos((prevTodos) => [...prevTodos, todo]);
+        })
+        .catch(() => {
+          setNotification(Errors.ErrorOnAdd);
+        })
+        .finally(() => {
+          clearInput();
+          setTempTodo(null);
+        });
+    }
+  }, []);
+
+  const handleDeleteTodo = useCallback((todoDel: Todo) => {
+    setDeletingId(prev => [...prev, todoDel.id]);
+
+    deleteTodo(todoDel.id)
+      .then(() => {
+        setTodos(prevTodos => (
+          prevTodos.filter(todo => todo.id !== todoDel.id)
+        ));
+      })
+      .catch(() => {
+        setNotification(Errors.ErrorOnDelete);
+      })
+      .finally(() => {
+        setDeletingId(prev => prev.filter(id => id !== todoDel.id));
+      });
+    clearNotification();
+  }, []);
+
+  const handleDeleteCompleted = useCallback(() => {
+    clearNotification();
+
+    completedTodoList.forEach(todo => handleDeleteTodo(todo));
+  }, [todos]);
+
+  const handleUpdate = useCallback(
+    (
+      updatingTodo: Todo,
+      newField: Partial<Todo>,
+    ) => {
+      setDeletingId(prev => [...prev, updatingTodo.id]);
+      updateTodo(updatingTodo.id, newField)
+        .then(updatedTodo => {
+          setTodos(prevTodos => (
+            prevTodos.map(todo => (
+              todo === updatingTodo
+                ? updatedTodo
+                : todo
+            ))
+          ));
+        })
+        .catch(() => setErrorNotification(Errors.ErrorOnUpdating))
+        .finally(() => {
+          setDeletingId(prev => prev.filter(id => id !== updatingTodo.id));
+        });
```

Comment:
Didn't get it, why do you first add Todo to `deletingId` then remove?

Reply:
I've made it to show the loader while new todo is adding.